File: javascript/alarm_rpc.js
#!/usr/bin/env node
"use strict";
// One MQTT alarm RPC; default is a read-only capabilities query.
const fs = require("node:fs");
const path = require("node:path");
const { randomInt } = require("node:crypto");
const { parseArgs } = require("node:util");
const { VdmCodec, VdmMqttClient, VdmTopics, parsePayloadFormat } = require("./sdk");

async function main() {
  const { values } = parseArgs({ options: {
    case: { type: "string" }, request: { type: "string" },
    "print-only": { type: "boolean", default: false }, listen: { type: "string", default: "0" },
  }});
  if (values.case && values.request) throw new Error("Choose --case or --request");
  const seconds = Number(values.listen);
  if (!Number.isFinite(seconds) || seconds < 0) throw new Error("--listen must be nonnegative seconds");
  const format = parsePayloadFormat(process.env.VDM_PAYLOAD_FORMAT || "protobuf");
  let request;
  if (values.request) request = JSON.parse(fs.readFileSync(values.request, "utf8"));
  else {
    const cases = JSON.parse(fs.readFileSync(path.join(__dirname, "../examples/alarms/requests.json"), "utf8"));
    const item = cases.find((item) => item.name === (values.case || "capabilities"));
    if (!item) throw new Error("Unknown case");
    request = { method: item.method, params: item[format] };
  }
  const reqId = randomInt(1, 2147483647);
  const encoded = new VdmCodec(format).encodeRpcRequest(request.method, request.params, reqId);
  console.log(JSON.stringify(request, null, 2));
  console.log(`format=${format} bytes=${encoded.payload.length} reqId=${reqId}`);
  if (values["print-only"]) return;
  if (!process.env.MQTT_HOST || !process.env.VDM_DEVICE_ID) throw new Error("Set MQTT_HOST and VDM_DEVICE_ID");
  const client = new VdmMqttClient({
    host: process.env.MQTT_HOST, port: Number(process.env.MQTT_PORT || "1883"),
    topics: VdmTopics.forDevice(process.env.VDM_DEVICE_ID), payloadFormat: format, qos: 1,
    username: process.env.MQTT_USERNAME, password: process.env.MQTT_PASSWORD,
  }, (message) => {
    if (["3A", "event"].includes(message.suffix)) console.log(message.suffix, JSON.stringify(message.data));
  }, (error) => console.error("DECODE_ERROR", error.message));
  try {
    await client.start();
    const response = await client.call(request.method, request.params, { reqId, timeoutMs: 30_000 });
    console.log("RESPONSE", JSON.stringify(response.data));
    if (seconds) await new Promise((resolve) => setTimeout(resolve, seconds * 1000));
  } finally {
    await client.stop();
  }
}
main().catch((error) => { console.error(error.message); process.exitCode = 1; });

File: python/alarm_rpc.py
#!/usr/bin/env python3
"""Run one MQTT alarm RPC. No arguments means read-only capabilities."""
import argparse
import json
import os
from pathlib import Path
import secrets
import time
from vdm_mqtt_sdk import (
    PayloadFormat,
    VdmCodec,
    VdmMqttClient,
    VdmMqttClientConfig,
    VdmTopics,
)


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    source = parser.add_mutually_exclusive_group()
    source.add_argument(
        "--case", default="capabilities", help="Name in examples/alarms/requests.json"
    )
    source.add_argument(
        "--request",
        type=Path,
        help="Your complete {method,params} JSON file for the selected format",
    )
    parser.add_argument(
        "--print-only",
        action="store_true",
        help="Validate encoding and print without connecting",
    )
    parser.add_argument(
        "--listen",
        type=float,
        default=0,
        help="Seconds to receive 3A/event after the RPC",
    )
    args = parser.parse_args()
    if args.listen < 0:
        parser.error("--listen must be nonnegative")
    profile = PayloadFormat.parse(os.getenv("VDM_PAYLOAD_FORMAT", "protobuf")).value
    if args.request:
        request = json.loads(args.request.read_text())
    else:
        cases = json.loads(
            (
                Path(__file__).resolve().parents[1] / "examples/alarms/requests.json"
            ).read_text()
        )
        case = next((item for item in cases if item["name"] == args.case), None)
        if case is None:
            parser.error("Unknown case: " + args.case)
        request = {"method": case["method"], "params": case[profile]}
    req_id = secrets.randbelow(2**31 - 1) + 1
    codec = VdmCodec(profile)
    payload, _ = codec.encode_rpc_request(request["method"], request["params"], req_id)
    print(json.dumps(request, ensure_ascii=False, indent=2))
    print(f"format={profile} bytes={len(payload)} reqId={req_id}")
    if args.print_only:
        return

    def on_message(message):
        if message.suffix in ("3A", "event"):
            # Do not call synchronous RPC from Paho's network callback.
            print(
                message.suffix,
                json.dumps(message.as_dict(), ensure_ascii=False),
                flush=True,
            )

    with VdmMqttClient(
        VdmMqttClientConfig(
            host=os.environ["MQTT_HOST"],
            port=int(os.getenv("MQTT_PORT", "1883")),
            topics=VdmTopics.for_device(os.environ["VDM_DEVICE_ID"]),
            payload_format=profile,
            username=os.getenv("MQTT_USERNAME"),
            password=os.getenv("MQTT_PASSWORD"),
        ),
        on_message=on_message,
        on_error=lambda error: print("DECODE_ERROR", error, flush=True),
    ) as client:
        response = client.call(
            request["method"], request["params"], req_id=req_id, timeout=30
        )
        print(
            "RESPONSE", json.dumps(response.as_dict(), ensure_ascii=False), flush=True
        )
        if args.listen:
            time.sleep(args.listen)


if __name__ == "__main__":
    main()
